source/common/event/dispatcher_impl.cc

```diff
@@ -74,7 +76,12 @@ DispatcherImpl::DispatcherImpl(const std::string& name, Api::Api& api,
       std::bind(&DispatcherImpl::updateApproximateMonotonicTime, this));
 }
 
-DispatcherImpl::~DispatcherImpl() { FatalErrorHandler::removeFatalErrorHandler(*this); }
+DispatcherImpl::~DispatcherImpl() {
+  ENVOY_LOG(debug, "destroying dispatcher {}", name_);
+  FatalErrorHandler::removeFatalErrorHandler(*this);
+  // TODO(lambdai): Resolve https://github.com/envoyproxy/envoy/issues/15072 and enable
+  // ASSERT(deletable_in_dispatcher_thread_.empty())
```

Comment:
Sure. But we should enforce the shutdown(). Otherwise, a benign deferred deleted could insert new dispatch thread local obj to dispatcher, and vice versa. In `~dispatcher` the obj list are destroyed instead of `clear` in shutdown. 

We don't see the above failure because the connection-host-cluster-sslctx destroy chain is straightforward.  IHMO dispatcher should call shutdown() to prevent itself

Reply:
I think it should be illegal to have objects deleted by deferred delete to schedule deferred deletions.  Of course, that's not enforced today.  You're uncovering a lot of issues in the dispatcher interface that will take some time to fully resolve.  It may be worth documenting them in separate issues.

Reply:
> I think it should be illegal to have objects deleted by deferred delete to schedule deferred deletions.

That is too good to be true. There might be cascading deletion. I encountered this in listener (deferred) deletion and the connection attached to that listener. I almost introduced the third deferred delete in between. 

I cannot resolve it for now but let me create a separate issue to track this

Reply:
I think this is for me to fix.  File an issue and send it my way.

Reply:
https://github.com/envoyproxy/envoy/issues/15111